core/store/src/trie/update.rs

```diff
@@ -72,7 +87,15 @@ impl TrieUpdate {
             }
         }
 
-        self.trie.get(&self.root, &key)
+        match &self.flat_state {
+            Some(flat_state) => match flat_state.get_ref(&key)? {
+                Some(ValueRef { hash, .. }) => {
+                    self.trie.storage.retrieve_raw_bytes(&hash).map(|bytes| Some(bytes.to_vec()))
+                }
+                None => Ok(None),
+            },
+            None => self.trie.get(&self.root, &key),
```

Comment:
So it is possible that there are some keys in the trie that are not in flat storage?  I thought flat storage would always contain the latest state i.e. it will have all keys that the trie has.

Reply:
No, this arm corresponds to the case "if TrieUpdate has FlatState object".
It has to be optional. When we process blocks, FS is necessary, but for view calls, state viewer, etc. we want to avoid using FS. In such cases `flat_state = None` and we always use Trie.
Also, TrieUpdate will have to be renamed to StateUpdate.

Reply:
Got it.  I think at some point it will be worth it for us to think about the types here so that "impossible" situations are not representable in the type system.  But let's leave that discussion for the future.